### PR TITLE
fix and refactor ApiClient and nextRows.

### PR DESCRIPTION
**Pull request by youngsofun**
1. ensure closed exactly once and close as early as possible.
2. use sessionStateRaw for compatibility.
3. unify retry policy.
4. allow QueryResponse.Schema to be nil.
5. one place to call NextQuery().
6. better code reuse to easy maintain. 
7. allow QueryResponse.Stats to be nil.
8. fix warnings.

### CODE REVIEW

**Comment by youngsofun** [comment]
cc @hantmac

File: connection.go [hunk 1]
```diff
@@ -28,82 +21,48 @@ type DatabendConn struct {
...
-				return emptyResult, nil
-			}
-		case resp := <-respCh:
-			b, err := json.Marshal(resp.Data)
-			if err != nil {
-				return emptyResult, err
-			}
-			_, _ = io.Copy(io.Discard, bytes.NewReader(b))
-		}
+	_, err := dc.rest.QuerySync(ctx, query, args)
```
Comment (hantmac) [inline connection.go:31]: We need to use `for-select` to wait the `QuerySync` until it end.

File: connection.go [hunk 2]
```diff
@@ -28,82 +21,48 @@ type DatabendConn struct {
...
-			if err != nil && (IsProxyErr(err) || strings.Contains(err.Error(), ProvisionWarehouseTimeout)) {
-				return true
-			}
-			return false
-		}),
-		retry.Delay(2*time.Second),
-		retry.Attempts(5),
-		retry.DelayType(retry.FixedDelay),
-	)
+func (dc *DatabendConn) query(ctx context.Context, query string, args ...driver.Value) (rows driver.Rows, err error) {
```
Comment (hantmac) [inline connection.go:38]: Do not remove the retry logic in `query`.

Comment (youngsofun) [reply connection.go:38]: retry is moved to seperate requests

File: connection.go (hunk 1 above)
Comment (youngsofun) [reply connection.go:31]: now waiting inside QuerySync, as it`s name implies.

**Review by hantmac — APPROVED** [review]
LGTM

**Pull request merged**